File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/infrastructure/repositories/infra/IQueryRepository.java
package com.oddly.ddd.infrastructure.repositories.infra;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

/**
 * Base interface for query repositories (read operations in CQRS).
 * Query repositories work with ReadEntities.
 * Returns ReadEntity directly to service (no BMO in read path).
 * 
 * @param <TReadEntity> The read entity type
 * @param <TId> The identifier type
 */
public interface IQueryRepository<TReadEntity, TId> {
    /**
     * Finds an entity by its identifier.
     * 
     * @param p_id The identifier to search for
     * @return Optional containing the entity if found, empty otherwise
     */
    CompletableFuture<Optional<TReadEntity>> findByIdAsync(TId p_id);

    /**
     * Lists all entities with pagination.
     * 
     * @param p_skip Number of records to skip
     * @param p_take Number of records to take
     * @return List of entities
     */
    CompletableFuture<List<TReadEntity>> listAsync(int p_skip, int p_take);

    /**
     * Counts the total number of entities.
     * 
     * @return Total count
     */
    CompletableFuture<Long> countAsync();
}

File: infrastructures/ddd/csharp/src/Infrastructure/Repositories/IQueryRepository.cs
using System.Collections.Generic;
using System.Threading.Tasks;

namespace OddlyDdd.Infrastructure.Repositories
{
    /// <summary>
    /// Base interface for query repositories (read operations in CQRS).
    /// Query repositories return ReadEntities directly to the service layer.
    /// No BMO mapping in the read path - ReadEntities are optimized views for queries.
    /// </summary>
    /// <typeparam name="TReadEntity">The read entity type</typeparam>
    /// <typeparam name="TId">The identifier type</typeparam>
    public interface IQueryRepository<TReadEntity, TId>
    {
        /// <summary>
        /// Finds a read entity by its identifier.
        /// </summary>
        /// <param name="p_id">The identifier to search for</param>
        /// <returns>The read entity if found, null otherwise</returns>
        Task<TReadEntity?> FindByIdAsync(TId p_id);

        /// <summary>
        /// Lists all read entities.
        /// </summary>
        /// <returns>A collection of read entities</returns>
        Task<IReadOnlyList<TReadEntity>> ListAllAsync();

        /// <summary>
        /// Lists read entities with pagination.
        /// </summary>
        /// <param name="p_skip">Number of items to skip</param>
        /// <param name="p_take">Number of items to take</param>
        /// <returns>A collection of read entities</returns>
        Task<IReadOnlyList<TReadEntity>> ListAsync(int p_skip, int p_take);

        /// <summary>
        /// Counts all read entities.
        /// </summary>
        /// <returns>The total count of entities</returns>
        Task<long> CountAsync();
    }
}
